### PR TITLE
Add tools/update-submodules.sh + a few other changes to how we build things

### LINKED ISSUE #108: Add global cleaning file
Would be nice to have some kind of global way of cleaning, including all the dependancies. This could be a `./clean.sh`, which recursively calls a `./clean.sh` for each solver (for example).

Also, do I have to "do something" when the vendored git submodules are updated? Wasn't sure.

This could be done via `git clean --force`, although that has the disadvantage it would nuke files people are editing, not just things that got built.

We could (in principle) even test our clean script cleans, by doing a full build, at the end cleaning, and checking something like `git status --ignored --ignore-submodules=none` comes back clean.

This is useful when some files don't get autorebuilt -- I don't think we want to start doing a full scan of dependancies at every build step, but also if someone does edit a file in one it might be annoying if they can't figure out how to make everything build again.

### PR DESCRIPTION
**Pull request by niklasdewally**
Closes #108 

@ChrisJefferson is this the kind of thing you had in mind?

### CODE REVIEW

File: clean.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+cd $(dirname "$0")
+
+# clear submodules
+bash ./etc/scripts/get_submodule_paths.sh | xargs rm -rf
```
Comment (ozgurakgun) [inline clean.sh:5]: I guess this is fine, bu could also do something like `git clean -dfx .` in each submodule so we don't fetch a fresh copy?

**Comment by niklasdewally** [comment]
@ozgurakgun I've done those changes.

I think git submodule sync then update covers all that could possibly change with submodules over time (i.e. what repo and branch it should point to, and updating it to new commits)?

**Comment by ChrisJefferson** [comment]
This doesn't remove solvers/chuffed/libwrapper.a and solvers/chuffed/wrapper.o . Not entirely sure who/what is responsable for that? can we teach `cargo clean` to do that? or do we need to run ./clean.sh from solvers/chuffed?

**Comment by ozgurakgun** [comment]
Here they say putting the build artefacts under OUT_DIR is the way to go: https://github.com/rust-lang/cargo/issues/2573

**Comment by ozgurakgun** [comment]
And this make me think configurable cargo clean isn't a thing: https://github.com/rust-lang/cargo/issues/572

**Comment by niklasdewally** [comment]
Found some typos in build.rs files! ( i think).

**Comment by niklasdewally** [comment]
> > Here they say putting the build artefacts under OUT_DIR is the way to go: [rust-lang/cargo#2573](https://github.com/rust-lang/cargo/issues/2573)
> 
> I'll open an issue for this so that Kieran can look at it when he has time (alongside the build stuff chris mentioned in the other PR). From the cargo book, it seems that it is heavily recommended to only use OUT_DIR in build scripts (probably for these reasons).
> 
> It looks like it should pass down as an env var. https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts

As I was in the build.rs files anyways, I have fixed this.

Also added the rerun-if-changed for the wrapper stuff as in #106.

I think the typo i fixed for this and minion would explain some issues Felix was having the other day - cargo build should've rebuilt the stuff in vendor but it didn't due to the typo, so we got a linking error.

**Comment by niklasdewally** [comment]
@Kieranoski702 As i've changed some chuffed stuff, would appreciate a review from you as well.

**Comment by niklasdewally** [comment]
> I think the typo i fixed for this and minion would explain some issues Felix was having the other day - cargo build should've rebuilt the stuff in vendor but it didn't due to the typo, so we got a linking error.

@ChrisJefferson @ozgurakgun  my understanding is with this fixed in [873d0d4](https://github.com/conjure-cp/conjure-oxide/pull/109/commits/873d0d4a13beefcbd432bd895815a58c802f3e89) we shouldn't need to clean build things on vendor update, but it might be nice to have anyways?

**Comment by niklasdewally** [comment]
> Here they say putting the build artefacts under OUT_DIR is the way to go: [rust-lang/cargo#2573](https://github.com/rust-lang/cargo/issues/2573)

Should we do this for solver build artifacts too (i.e the stuff currently in vendor/build), or is what we have good enough for now?

File: solvers/minion/build.sh [hunk 2]
```diff
@@ -2,20 +2,17 @@
 
 SCRIPT_DIR=$(dirname "$0")
 
-git submodule init
-git submodule update
+git submodule init -- vendor
+git submodule sync -- vendor 
+git submodule update --init --recursive -- vendor
+
 cd "$SCRIPT_DIR"
```
Comment (ozgurakgun) [inline solvers/minion/build.sh:9]: it looks like this cd wants to be before the git commands, am I missing something?

Comment (niklasdewally) [reply solvers/minion/build.sh:9]: thank you - thats a typo on my part.

File: solvers/chuffed/build.sh [hunk 3]
```diff
@@ -1,17 +1,26 @@
 #!/bin/bash
 
-SCRIPT_DIR=$(dirname "$0")
+SCRIPT_DIR=$(realpath $(dirname "$0"))
```
Comment (ozgurakgun) [inline solvers/chuffed/build.sh:3]: why the realpath here and not in the minion one? also same question about location of cd.

File: etc/scripts/get_submodule_paths.sh [hunk 4]
```diff
@@ -6,7 +6,7 @@
 # these paths are relative to the git repository.
 
 # go to top level of git repo
-cd $(git rev-parse --show-toplevel)
+cd $(git rev-parse --show-superproject-working-tree --show-toplevel | head -n 1)
```
Comment (ozgurakgun) [inline etc/scripts/get_submodule_paths.sh:9]: I learned a new arg to git today! `--show-superproject-working-tree`

**Comment by ozgurakgun** [comment]
Sorry I added comments one-by-one instead of a combined review. I always think I'll just say this one thing, and end up saying more things...

If we managed to make cargo clean/build do the right thing I vote for removing clean.sh.

File: solvers/chuffed/build.sh (hunk 3 above)
Comment (niklasdewally) [reply solvers/chuffed/build.sh:3]: this has already been fixed in a WIP commit - the short answer is that I am doing more stuff with minion atm.

**Comment by niklasdewally** [comment]
> Sorry I added comments one-by-one instead of a combined review. I always think I'll just say this one thing, and end up saying more things...
> 
> If we managed to make cargo clean/build do the right thing I vote for removing clean.sh.

I have made it do the right thing, I think.

I'll leave clean.sh for now for Chris and Kieran to review, just in case I've missed some C++ stuff somewhere.

**Review by ozgurakgun — APPROVED** [review]
LGTM as far as anyone can review shell scripts by reading...

**Comment by ChrisJefferson** [comment]
cargo clean now gets rid of all the C++ files, as far as I can see. I think having a global clean.sh to reset all the submodules is still nice (as it does seem to require multiple lines), but cargo clean will now do most of the work.

**Comment by ozgurakgun** [comment]
So there are 3 parts to the clean script:

- `git submodule sync --recursive` my understanding is that this updates the submodules to use the up to date URLs from the .gitmodules file
- `git submodule update --init --recursive` this updates the submodules to their latest version
- ` git clean -dfx .` on submodules to remove untracked files etc

So this is less to do with cleaning and more to do with updating the submodules, right? Should we rename it to `update-submodules.sh`? Or even `etc/build/update-submodules.sh` so it's not sitting at the top?

We should also add a link to the readme file pointing to the [contributing page](https://github.com/conjure-cp/conjure-oxide/wiki/Contributing) which I see @niklasdewally updated yesterday to mention clean.sh.

**Comment by niklasdewally** [comment]
> So there are 3 parts to the clean script:
> 
> - `git submodule sync --recursive` my understanding is that this updates the submodules to use the up to date URLs from the .gitmodules file
> - `git submodule update --init --recursive` this updates the submodules to their latest version
> - ` git clean -dfx .` on submodules to remove untracked files etc
> 
> So this is less to do with cleaning and more to do with updating the submodules, right? Should we rename it to `update-submodules.sh`? Or even `etc/build/update-submodules.sh` so it's not sitting at the top?
> 
> We should also add a link to the readme file pointing to the [contributing page](https://github.com/conjure-cp/conjure-oxide/wiki/Contributing) which I see @niklasdewally updated yesterday to mention clean.sh.

etc/scripts/update-submodules.sh sounds good


Would tools be a better name for the etc directory? It might be nicer if we have

tools/foo.sh instead of etc/scripts/foo.sh from an ergonomic perspective

**Comment by ozgurakgun** [comment]
I am happy with that since tools exists already.

**Comment by niklasdewally** [comment]
> I am happy with that since tools exists already.

Oh I didn't notice that! Consistent directory structure either way would be good and tools actually says what it does so.

**Comment by niklasdewally** [comment]
I've just added some stuff to the wiki page about this and installation generally as well

**Pull request merged**